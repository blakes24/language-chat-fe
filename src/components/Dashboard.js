import { useEffect, useState } from "react";
import UserList from "./UserList";
import Container from "@material-ui/core/Container";
import ChatApi from "../helpers/api";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles({
  root: {
    margin: "1rem",
  }
});

function Dashboard() {
  const classes = useStyles();
  const [users, setUsers] = useState([]);

  useEffect(() => {
    async function getUsers() {
      const res = await ChatApi.getAllUsers();
      setUsers(res);
    }
    getUsers();
  }, []);

  return (
    <Container>
      <Typography variant="h6" component="h1" align='center' className={classes.root}>
        Find a partner and start chatting!
      </Typography>
      {users.length && <UserList users={users} />}
    </Container>
  );
}

export default Dashboard;
