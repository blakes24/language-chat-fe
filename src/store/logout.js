import { createAction } from "@reduxjs/toolkit";

export const logoutUser = createAction("LOGOUT_USER");
