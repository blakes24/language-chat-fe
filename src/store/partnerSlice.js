import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import ChatApi from "../helpers/api";

export const fetchPartners = createAsyncThunk(
  "partners/fetchPartners",
  async (userId, { rejectWithValue }) => {
    try {
      const response = await ChatApi.getPartners(userId);
      return response;
    } catch (err) {
      return rejectWithValue(err);
    }
  }
);

export const partnerSlice = createSlice({
  name: "partners",
  initialState: { items: [], loading: "idle", error: null },
  reducers: {},
  extraReducers: {
    [fetchPartners.pending]: (state, action) => {
      if (state.loading === "idle") {
        state.error = null;
        state.loading = "pending";
      }
    },
    [fetchPartners.fulfilled]: (state, action) => {
      if (state.loading === "pending") {
        state.loading = "idle";
        state.items = action.payload;
      }
    },
    [fetchPartners.rejected]: (state, action) => {
      if (state.loading === "pending") {
        state.loading = "idle";
        state.error = action.payload;
      }
    },
  },
});

export default partnerSlice.reducer;
