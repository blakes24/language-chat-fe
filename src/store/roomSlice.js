import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import ChatApi from "../helpers/api";

// data = { userId, partnerId }
export const fetchRooms = createAsyncThunk(
  "rooms/fetchRooms",
  async (data, { rejectWithValue }) => {
    try {
      const response = await ChatApi.getRooms(data);
      return response.rooms;
    } catch (err) {
      return rejectWithValue(err);
    }
  }
);

export const roomSlice = createSlice({
  name: "rooms",
  initialState: { items: [], current: null, loading: "idle", error: null },
  reducers: {
    setCurrentRoom(state, action) {
      state.current = action.payload;
    },
    addRoom(state, action) {
      state.items.push(action.payload);
    },
  },
  extraReducers: {
    [fetchRooms.pending]: (state, action) => {
      if (state.loading === "idle") {
        state.error = null;
        state.loading = "pending";
      }
    },
    [fetchRooms.fulfilled]: (state, action) => {
      if (state.loading === "pending") {
        state.loading = "idle";
        state.items = action.payload;
      }
    },
    [fetchRooms.rejected]: (state, action) => {
      if (state.loading === "pending") {
        state.loading = "idle";
        state.items = [];
        state.error = action.payload;
      }
    },
  },
});

export const { setCurrentRoom, addRoom } = roomSlice.actions;

export default roomSlice.reducer;
