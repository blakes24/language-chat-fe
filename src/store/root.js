import { configureStore } from "@reduxjs/toolkit";
import usersReducer from "./usersSlice";
import roomReducer from "./roomSlice";
import messageReducer from "./messageSlice";

export default configureStore({
  reducer: {
    users: usersReducer,
    rooms: roomReducer,
    messages: messageReducer,
  },
});
