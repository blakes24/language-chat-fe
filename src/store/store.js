import { configureStore, getDefaultMiddleware } from "@reduxjs/toolkit";
import storage from "redux-persist/lib/storage";
import {
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from "redux-persist";
import rootReducer from "./root";

const persistConfig = {
  key: "root",
  storage,
  whitelist: ["users"],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  devTools: process.env.NODE_ENV !== "production",
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
});

export default store;
