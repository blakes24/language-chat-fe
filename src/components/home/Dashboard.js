import { useEffect, useState } from "react";
import UserList from "../user/UserList";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";
import InputLabel from "@material-ui/core/InputLabel";
import MenuItem from "@material-ui/core/MenuItem";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import { useSelector, useDispatch } from "react-redux";
import { fetchAllUsers } from "../../store/usersSlice";
import Loading from "../Loading";

const useStyles = makeStyles({
  root: {
    marginTop: "1.5rem",
    marginBottom: ".5rem",
  },
  formControl: {
    marginBottom: ".5rem",
    minWidth: 240,
  },
  formContainer: {
    display: "flex",
    justifyContent: "center",
  },
  err: {
    textAlign: "center",
  },
});

function Dashboard() {
  const classes = useStyles();
  const [filter, setFilter] = useState("");
  const users = useSelector((state) => state.users.items);
  const loading = useSelector((state) => state.users.loading);
  const error = useSelector((state) => state.users.error);
  const dispatch = useDispatch();

  const languages = [
    { label: "Any", value: "" },
    { label: "Arabic", value: "ar" },
    { label: "Chinese", value: "zh" },
    { label: "English", value: "en" },
    { label: "French", value: "fr" },
    { label: "German", value: "de" },
    { label: "Hindi", value: "hi" },
    { label: "Indonesian", value: "id" },
    { label: "Italian", value: "it" },
    { label: "Japanese", value: "ja" },
    { label: "Korean", value: "ko" },
    { label: "Portuguese", value: "pt" },
    { label: "Russian", value: "ru" },
    { label: "Spanish", value: "es" },
  ];

  const handleChange = (event) => {
    setFilter(event.target.value);
  };

  useEffect(() => {
    function getUsers() {
      try {
        dispatch(fetchAllUsers(filter));
      } catch (err) {
        console.error(err);
      }
    }
    getUsers();
  }, [filter, dispatch]);

  return (
    <Container>
      <Typography
        variant="h6"
        component="h1"
        align="center"
        className={classes.root}
      >
        Find a partner and start chatting!
      </Typography>
      <Container className={classes.formContainer}>
        <FormControl
          variant="outlined"
          margin="dense"
          className={classes.formControl}
        >
          <InputLabel id="language">Show users who speak...</InputLabel>
          <Select
            labelId="language"
            id="language"
            value={filter}
            onChange={handleChange}
            label="Show users who speak..."
          >
            {languages.map((lang) => (
              <MenuItem value={lang.value} key={lang.value || 1}>
                {lang.label}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      </Container>
      {loading === "pending" && <Loading solid />}
      {error &&
        error.map((err, idx) => (
          <p className={classes.err} key={idx}>
            {err}
          </p>
        ))}
      {users.length > 0 && <UserList users={users} />}
    </Container>
  );
}

export default Dashboard;
