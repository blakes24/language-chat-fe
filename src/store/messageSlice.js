import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import ChatApi from "../helpers/api";
import { logoutUser } from "./logout";

export const fetchMessages = createAsyncThunk(
  "messages/fetchMessages",
  async (roomId, { rejectWithValue, dispatch }) => {
    try {
      const response = await ChatApi.getMessages(roomId);
      return response;
    } catch (err) {
      if (err?.[0] === "Invalid token") {
        dispatch(logoutUser());
      } else {
        return rejectWithValue(err);
      }
    }
  }
);

export const messageSlice = createSlice({
  name: "messages",
  initialState: { items: [], loading: "idle", error: null },
  reducers: {
    addMessage(state, action) {
      state.items.push(action.payload);
    },
  },
  extraReducers: {
    [fetchMessages.pending]: (state, action) => {
      if (state.loading === "idle") {
        state.error = null;
        state.loading = "pending";
      }
    },
    [fetchMessages.fulfilled]: (state, action) => {
      if (state.loading === "pending") {
        state.loading = "idle";
        state.items = action.payload;
      }
    },
    [fetchMessages.rejected]: (state, action) => {
      if (state.loading === "pending") {
        state.loading = "idle";
        state.items = [];
        state.error = action.payload;
      }
    },
  },
});

export const { addMessage } = messageSlice.actions;

export default messageSlice.reducer;
