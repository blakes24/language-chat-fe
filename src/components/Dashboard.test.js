import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import Dashboard from "./Dashboard";
import store from "../store/root";
import { Provider } from "react-redux";

test("renders without crashing", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <Dashboard />
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Dashboard />
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
