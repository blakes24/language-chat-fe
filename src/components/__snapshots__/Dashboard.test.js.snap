// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <h1
      class="MuiTypography-root makeStyles-root-2 MuiTypography-h6 MuiTypography-alignCenter"
    >
      Find a partner and start chatting!
    </h1>
    0
  </div>
</DocumentFragment>
`;
